lib/astutils.cpp

```diff
@@ -789,11 +789,11 @@ bool extractForLoopValues(const Token *forToken,
     *knownInitValue = initExpr->astOperand2()->hasKnownIntValue();
     *initValue = minInitValue.empty() ? 0 : minInitValue.front();
     *partialCond = Token::Match(condExpr, "%oror%|&&");
-    visitAstNodes(condExpr, [varid, &condExpr](const Token *tok) {
+    visitAstNodes(condExpr, [varid, &condExpr, partialCond](const Token *tok) {
         if (Token::Match(tok, "%oror%|&&"))
             return ChildrenToVisit::op1_and_op2;
         if (Token::Match(tok, "<|<=") && tok->isBinaryOp() && tok->astOperand1()->varId() == *varid && tok->astOperand2()->hasKnownIntValue()) {
-            if (Token::Match(condExpr, "%oror%|&&") || tok->astOperand2()->getKnownIntValue() < condExpr->astOperand2()->getKnownIntValue())
+            if (*partialCond || tok->astOperand2()->getKnownIntValue() < condExpr->astOperand2()->getKnownIntValue())
```

Comment:
This is totally bogus since `condExpr` is assigned on the next line so it might be different on each call. Will revert. We could perform the lighter check first though.